src/preprocess/links.rs

```diff
@@ -89,19 +89,26 @@ where
     let mut replaced = String::new();
 
     for link in find_links(s) {
-        replaced.push_str(&s[previous_end_index..link.start_index]);
+        let before_link = &s[previous_end_index..link.start_index];
+        let offset = if let Some(i) = before_link.rfind('\n') {
+            &before_link[i + 1..]
+        } else {
+            ""
+        };
+        replaced.push_str(before_link);
 
         match link.render_with_path(&path, chapter_title) {
             Ok(new_content) => {
                 if depth < MAX_LINK_NESTED_DEPTH {
                     if let Some(rel_path) = link.link_type.relative_path(path) {
-                        replaced.push_str(&replace_all(
-                            &new_content,
-                            rel_path,
-                            source,
-                            depth + 1,
-                            chapter_title,
-                        ));
+                        let v =
+                            replace_all(&new_content, rel_path, source, depth + 1, chapter_title);
+                        let lines = v.split('\n').into_iter().collect::<Vec<&str>>();
```

Comment:
Since you're iterating over the lines one-by-one, you could iterate directly over the result of `split`. So you would save the `collect()` call be a tiny bit more efficient.